src/array.c

```diff
@@ -483,6 +483,12 @@ JL_DLLEXPORT jl_array_t *jl_pchar_to_array(const char *str, size_t len)
 JL_DLLEXPORT jl_value_t *jl_array_to_string(jl_array_t *a)
 {
     size_t len = jl_array_len(a);
+    if (len == 0) {
+        // this may seem like purely an optimization (which it also is), but it
+        // also ensures that calling `String(a)` doesn't corrupt a previous
```

Comment:
Just curious, but why/how exactly does a previous string get corrupted w/o this? The segfault seems to be gc-related; something about the original string now being "shared" between two strings?

Reply:
We truncate the array here (so that future uses will get a fresh data object), but that ends up truncating the string later, if it is still the data pointer and we call this immediately. See the added tests.